### ISSUE TITLE
Support more extensions

### ISSUE THREAD

**Issue by crazywhalecc** [crazywhalecc/static-php-cli#152]
## Notice

Here is a list of supported extensions: <https://static-php.dev/en/guide/extensions.html> .

If you need a new extension to be added, please [create a new issue](https://github.com/crazywhalecc/static-php-cli/issues/new) and attach a link to the extension. 

Listed below are extensions that have been proven to be unsupported.

## Unsupported Extensions

- pcov: Zend module, dynamic link only
- [tonyenc](https://gitee.com/lfveeker/tonyenc): Not usable #203 (temporarily closed)
- ardillo/libui: #317 
- vips: #386 

<details>

<summary>Old support list (deprecated, no longer updated)</summary>

## Linux & macOS

- [ ] [rdkafka](https://github.com/crazywhalecc/static-php-cli/issues/70) #256
- [ ] [pdo_snowflake](https://github.com/snowflakedb/pdo_snowflake)
- [x] ldap #193 
- [ ] [decimal](https://github.com/php-decimal/ext-decimal)
- [X] grpc
- [x] #202
- [ ] pcov (Zend module, dynamic link only)
- [X] xsl
- [x] [password-argon2](https://github.com/crazywhalecc/static-php-cli/pull/64)
- [x] [snappy](https://github.com/crazywhalecc/static-php-cli/pull/157)
- [ ] [tonyenc](https://gitee.com/lfveeker/tonyenc) #203 (temporarily closed)
- [x] gettext #339 
- [x] uv #237
- [x] [rar](https://pecl.php.net/package/rar) #260 
- [x] tidy #271
- [x] sqlsrv #291 
- [x] igbinary
- [ ] ardillo/libui #317
- [x] uuid #362 
- [x] amqp #366 
- [x] pdo_sqlsrv #374 
- [ ] vips #386 
- [x] simdjson #390 
- [x] yac #393 
- [x] ds #395 #401 
- [x] msgpack
- [x] gmssl 

## Windows

- [x] #429 
- [x] #435 
- [X] apcu #389 #488 
- [X] pdo, sqlsrv #389 #488 
- [X] yaml #389 #488 
- [X] rar #389 #488 
- [X] zip #389 #488 
- [x] msgpack
- [x] gmssl 

## FreeBSD

No.

Since FreeBSD lacks CI support on GitHub and it is troublesome to test on my own computer, I will temporarily put the update of FreeBSD on hold.

If someone can provide FreeBSD CI or answer questions related to FreeBSD compilation, I will be very grateful and restart the official maintenance of FreeBSD.

</details>

**Comment by stloyd** [comment]
Aren't those extensions already covered?
- [imap](https://github.com/crazywhalecc/static-php-cli/blob/main/config/ext.json#L133),
- [ldap](https://github.com/crazywhalecc/static-php-cli/blob/main/config/ext.json#L154)

**Comment by DubbleClick** [comment]
I think the old links that are mentioned in [git.alpinelinux.org](https://git.alpinelinux.org/aports/tree/main/imap/APKBUILD) are dead, but the sources can be found at https://github.com/uw-imap/imap

**Comment by crazywhalecc** [comment]
@DubbleClick sysvsem is supported now, but it seems I forgot add it in documentation. I'll add it later.

ldap is planned but I haven't started yet.

**Comment by DubbleClick** [comment]
I was able to compile php with ldap just following the documentation. Imap is a whole other story, had no luck with it on rhel 9.2. I hope I'll find the time to create a PR for ldap next week.

**Comment by crazywhalecc** [comment]
> I was able to compile php with ldap just following the documentation. Imap is a whole other story, had no luck with it on rhel 9.2. I hope I'll find the time to create a PR for ldap next week.

It would be great if you could improve new extensions! 

And I noticed that you are using RHEL, I'm not familiar with it. It might be more helpful if someone could first improve the `doctor` check items for RHEL and CentOS environments.

**Comment by DubbleClick** [comment]
Yeah I've noticed that there was no musl auto-fix for rhel available. I wasn't able to find a rpm for it and therefore had to manually build it and add the directory to ld_flags.

Will do again with a fresh VM and document the necessary comments. I shound find time next week.

**Comment by DubbleClick** [comment]
I've had more of a play with imap and got it to compile (dynamically so far, with openssl3, not 1.1.1), but even that requires a lot of patches both in the form of .patch files as well as replacing contents in files that no .patch file is available for. This project doesn't have any logic to apply patch files yet.

It also needs multiple services to be running on the system and a (self signed is fine) certificate. It's a complete mess that requires decisions to be taken to integrate it here that I'm not in a position to make.

I would very much appreciate if someone could tackle it, because it's a very necessary extension with no real replacement (there's laminas mail to send mail with limited functionality, an abandoned php-imap2 that doesn't work with php 8, and nothing else). But you should expect it to be quite a lot of work. :/

**Comment by crazywhalecc** [comment]
@DubbleClick 

> but even that requires a lot of patches both in the form of .patch files as well as replacing contents in files that no .patch file is available for. This project doesn't have any logic to apply patch files yet.

I understand. I tried the patch code in my previous implementation of libmemcached, where I originally planned to put the patch code in the `src/globals/patch/` directory, like this. But in the end, I directly used homebrew's build logic and independent warehouse, and did not choose patch because it has not been updated for a long time: https://github.com/crazywhalecc/libmemcached-macos

<img width="353" alt="image" src="https://github.com/crazywhalecc/static-php-cli/assets/20330940/721ecddb-c5dc-4df1-9c7a-add411fe0621">

> It also needs multiple services to be running on the system and a (self signed is fine) certificate. It's a complete mess that requires decisions to be taken to integrate it here that I'm not in a position to make.

From the results of my current local testing, it requires at least kerberos, and I'm currently stuck on compiling kerberos. 

Under the existing architecture, it may not be a good idea to use pre-installed dynamic libraries or packages on the system, because this will limit the environment in which spc can run.

**Comment by boonkerz** [comment]
Currently i work on sdl 

cli and micro with static sdl3 lib currently only on windows

<img width="1147" alt="image" src="https://github.com/crazywhalecc/static-php-cli/assets/277321/444e7444-b2cc-4dd3-9ee4-e8170108f052">

**Comment by jorgeluiz1586** [comment]
Hello, could you include the Openswoole extension to be tested and included?

I saw that there is Swoole but I need the gRPC client and gRPC server of Openswoole, I believe it is compatible since Openswoole is a fork of Swoole with more features.

**Comment by crazywhalecc** [comment]
@jorgeluiz1586 It should not be difficult to support openswoole.

> But I am not sure if openswoole has a similar hook bug as swoole (phpmicro+phar cannot be used in PROCESS mode).

**Comment by weijer** [comment]
Need to support ：msgpack

**Comment by crazywhalecc** [comment]
@weijer Please create an issue and we can start this with progress.

**Comment by mrlovables** [comment]
Need to support ：GmSSL-PHP,

**Comment by crazywhalecc** [comment]
@mrlovables @weijer Now static-php supported gmssl and msgpack in 2.3.4 .

**Comment by mrlovables** [comment]
> @mrlovables @weijer Now static-php supported gmssl and msgpack in 2.3.4 .现在 static-php 在 2.3.4 中支持 gmssl 和 msgpack 。

thanks

**Comment by kdrmlhcn** [comment]
I was able to get a build with the "intl" extension for PHP 8.3 on MacOS, but I couldn't do it for Windows. I couldn't find a solution either. What do you recommend for this? @crazywhalecc

**Comment by rcmcdonald91** [comment]
Regarding FreeBSD CI : https://cirrus-ci.org/

**Comment by crazywhalecc** [comment]
> I was able to get a build with the "intl" extension for PHP 8.3 on MacOS, but I couldn't do it for Windows. I couldn't find a solution either. What do you recommend for this? @crazywhalecc

@kdrmlhcn For Windows, you can build a static php.exe with the same PHP version, same VS version, and same thread safety as windows.php.net, and then use php.ini to dynamically load the intl extension.

**Comment by crazywhalecc** [comment]
@rcmcdonald91 Thank you very much for the information. At the moment I'm more looking for a way to integrate with GHA, so it would be great if you could provide more help with CI.

**Comment by zengbo** [comment]
Need to support:  gnupg and grpc
